File: app/src/main/java/com/evaluation/adapter/CustomListAdapter.java
package com.evaluation.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.evaluation.command.ICommand;
import com.evaluation.countrylist.CountryQuery;
import com.evaluation.countrylist.R;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CustomListAdapter extends RecyclerView.Adapter<CustomListAdapter.ListAdapterHolder> {

    private final List<CountryQuery.Country> mCountriesList;
    private final ICommand<CountryQuery.Country> mClickCommand;


    public CustomListAdapter(List<CountryQuery.Country> countriesList, ICommand<CountryQuery.Country> clickCommand) {
        this.mCountriesList = countriesList;
        this.mClickCommand = clickCommand;
    }

    @NotNull
    @Override
    public ListAdapterHolder onCreateViewHolder(ViewGroup viewGroup, int position) {
        return new ListAdapterHolder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(ListAdapterHolder countriesListAdapterHolder, int position) {
        countriesListAdapterHolder.bind(getItem(position), mClickCommand);
    }


    private CountryQuery.Country getItem(int position) {
        return mCountriesList.get(position);
    }

    @Override
    public int getItemCount() {
        return mCountriesList.size();
    }

    static class ListAdapterHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.name)
        TextView titleView;

        ListAdapterHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }

        void bind(final CountryQuery.Country userList, ICommand<CountryQuery.Country> clickCommand) {

            itemView.setOnClickListener(v -> clickCommand.execute(userList));

            titleView.setText(userList.name());
        }
    }
}

File: app/src/main/java/com/evaluation/adapter/CustomListAdapter.kt
package com.evaluation.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.evaluation.adapter.CustomListAdapter.ListAdapterHolder
import com.evaluation.command.ICommand
import com.evaluation.countrylist.CountryQuery.Country
import com.evaluation.countrylist.R

class CustomListAdapter(private val mCountriesList: List<Country>, private val clickCommand: ICommand<Country>) : RecyclerView.Adapter<ListAdapterHolder>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, position: Int): ListAdapterHolder =
        ListAdapterHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.list_item, viewGroup, false))

    override fun onBindViewHolder(countriesListAdapterHolder: ListAdapterHolder, position: Int) {
        countriesListAdapterHolder.bind(getItem(position), clickCommand)
    }

    private fun getItem(position: Int): Country {
        return mCountriesList[position]
    }

    override fun getItemCount(): Int {
        return mCountriesList.size
    }

    class ListAdapterHolder(view: View?) : RecyclerView.ViewHolder(view!!) {

        @BindView(R.id.name)
        lateinit var titleView: TextView

        init {
            ButterKnife.bind(this, view!!)
        }

        fun bind(userList: Country, clickCommand: ICommand<Country>) {
            itemView.setOnClickListener { clickCommand.execute(userList) }
            titleView.text = userList.name()
        }
    }
}